### ISSUE TITLE
ACCESSIBILITY: Settings editor almost impossible to use with screen readers

### ISSUE THREAD

**Issue by zersiax** [Microsoft/vscode#50968]
- VSCode Version: 1.23
- OS Version: Windows 10 1803, but reproduceable on any version

Steps to Reproduce:

1. Open the settings pane using ctrl+,.
2. Type in the editor with a screen reader enabled, I tested with NVDA.

Expected result:
Screen reader does not interrupt, only reads suggestions for autocompletion.

Perceived result:
Screen reader reads "total x settings' where x is the number of available settings. On, every, keypress. That is, to put it lightly, extremely annoying, please fix.
Does this issue occur when all extensions are disabled?: Yes

**Comment by roblourens** [comment]
It was specifically changed to do that in https://github.com/Microsoft/vscode/issues/19143. Do you expect it to ever read "x settings found", like after a pause when typing? Or never?

**Comment by zersiax** [comment]
I must admit I'm a little vague on how exactly this is to be used to begin with. I see two completely seperate flows:
- You type in the editor itself and it would behave like any other code editor, giving you autocomplete suggestions as you go. For this to work, you need at least part of the settings name in order to autocomplete it. In this case, I would NOT want to hear 'total x settings', as I'm not actually doing a search, I'm just typing in my .json directly. I have mixed results with this approach, often I just get no autocomplete suggestions at all when typing in the settings.json file and I'm not sure what governs that.
- You type in the search field to search for a setting and ...something happens. I honestly have no idea where the filtered settings go, how to navigate through them, activate them to be added to the .json file etc. It doesn't seem to be going to a treeview like search results from a file search, so I'm not sure where to actually find my search results. As for the issue you linked, I don't really see the use of knowing how many search results I've found if I can't see what those results are, and if I do I find the use of such an announcement limited at best ... @isidorn was that issue raised after someone gave you guys feedback about it?

**Comment by roblourens** [comment]
That's right, if you type in the editor it should give suggestions like any other editor. In your original post you said you are typing in the editor, but I thought you meant you're typing in the search box, because that gives the behavior you describe.

When I type in the editor, the screenreader seems to act the same as it does in other editors - it reads details about the selected suggestion and doesn't read "total x settings".

Below the search field is a split editor. The search results are filtered on the left side of the split editor, and they can be copied to the right side. When I type in the search field, I hear what you describe, it reads "total x settings" for every keypress.

**Comment by zersiax** [comment]
I think we might be getting close to the issue. Let me describe to you what
I see:
In regards to dit fields on the page when the settings.json is open, I see
three edit fields. The first field is the search field. Typing in this
field indeed tells me how many settings there are.
The second is labeled as 'default settings' and is a read-only edit field.
I cannot type in this field, obviously.
The third, and final, edit field is an editor labeled settings.json, which
also echoes the "total x settings" on every letter and does not give me
suggestions for autocompletion.
Could it be that I'm typing in the wrong side of the split editor? I only
see one editor, no idea how to move from left to right.

2018-06-02 20:10 GMT+02:00 Rob Lourens <notifications@github.com>:

**Comment by roblourens** [comment]
I also tried this with NVDA. When I type in the final settings.json edit field, I hear it read the first suggestion in the suggestion popup, and I don't hear the number of setting matches.

Is it possible that you are typing in the settings search field, when you think you are typing in the setting.json editor? Can you tell me how you are switching between the three edit fields?

**Comment by zersiax** [comment]
I use e and shift+e to navigate between edit fields on the page. THESE ARE
NVDA hotkeys. I do encounter the behavior you describe at times, but I also
encounter the situation as described above and I'm not sure what the
deciding factor is.

2018-06-04 20:07 GMT+02:00 Rob Lourens <notifications@github.com>:

**Comment by roblourens** [comment]
I can't get that shortcut to work, have you customized it? I don't see it listed on this page: https://dequeuniversity.com/screenreaders/nvda-keyboard-shortcuts

**Comment by zersiax** [comment]
It's similar to f moving from form field to form field, except it doesn't target buttons, checkboxes etc. I don't think it's very common.
I've done some more testing, and WHEN I get the suggestions to show up in the main editor for settings.json, they are followed by the "total x settings" message. If you're able to hangouts/skype with me I can perhaps show this off.

**Comment by roblourens** [comment]
I actually am able to repro this now! This was just my lack of familiarity with NVDA. It's definitely a bug.

**Comment by zersiax** [comment]
awesome :) I'd love to see it fixed, since it makes the settings experience rather tedious. There's a few more things with that window ...I don't think you can currently read the default settings read-only edit field with NVDA, forcing you to copy-paste it into another editor for viewing/editing. I also haven't gotten search to work properly ...as in ...I can search for stuff, but the way to actually use the found results in settings.json from just the keyboard escapes me and NVDA doesn't seem to give any hints on how to accomplish this either. Can you check if you can get it done without using a mouse at all?

**Comment by roblourens** [comment]
I think the unnecessary 'total x settings' reading should be fixed in the next Insiders build.

If I highlight the text on the default settings side, NVDA reads it out. Is that not working for you?

I'll admit that search is awkward to use - if you type in the search bar, it will filter down the list of default settings on the left. Then they can be copy/pasted to the right. It may be easier to just type on the right and rely on intellisense, which should be much less annoying with this fix.

We are also working on a new settings editor which should be easier to make accessible than this multi-editor thing. It's buggy right now. But once it's more stable, I'd really like to get your opinion on how usable it is, if that's ok with you

**Comment by zersiax** [comment]
Just one final thing, did you highlight using the keyboard or the mouse? :-) I can arrow through the settings, but there's apparently some menu-driven way of viewing the possible options for a given settings. Selecting one of those options copies the setting to the right, edited and all. That part I haven't been able to trigger from the keyboard, and just typing the setting is how I set them now. However, for that to work you need to at least know part of the setting name, unless you want to read and memorize all the default settings :)

**Comment by roblourens** [comment]
I highlighted the line using the keyboard, and it reads the line.

There is a button that shows up next to a setting when the cursor is in its line. But it's not even in the tab order so I didn't mention it. When clicked, it shows a context menu with the possible values of a setting if it's a dropdown. For other types of settings, it just says "Copy to setting" and copies the setting to the right side of the editor.

**Comment by zersiax** [comment]
Right, that button is what I was referring to. There's currently no
keyboard-only equivalent to bring up that menu if I'm not mistaken.

Op di 12 jun. 2018 om 00:12 schreef Rob Lourens <notifications@github.com>:
